### PR TITLE
feat: Subscription Settings Page Created

### LINKED ISSUE #549: Create a subscription settings page
## Description
As a user,
I need be able to see a settings page,
so that I can select the kinds of notifications that will be sent to me.

Following options should be there in the settings: 

1. Community Updates
a. Upcoming Events for Communities I’m a member of (eg:- weekly digest that includes the upcoming events for multiple communities)
b. Updates from Community Interactions(eg:- weekly digest that includes stat changes of the community)

2. Meetup Updates for RSVP’ed Meetups
a. Receive updates on Location Change
b. Receive updates on Date Change
c. Receive updates on Time Changed
d. Reminders for the event (1 day prior, 1 hour prior)

## Mocks
In progress. Issue #544 

### Update [Required]
- [ ] Create the UI and the necessary Django views and models for making the feature functional
- [ ] Write necessary tests 

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

## Estimation
2-3 days

### PR DESCRIPTION
**Pull request by ritwickraj78**
### Description
Include a summary of the change and relevant motivation/context. List any dependencies that are required for this change.

Fixes #549 

### Type of Change:
**Delete irrelevant options.**

- Code
- User Interface

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Describe the tests you ran to verify your changes. Provide instructions or GIFs so we can reproduce. List any relevant details for your test.
Screenshot from 2020-07-16 14-30-05
Screenshot from 2020-07-16 14-29-48

### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: systers_portal/users/models.py [hunk 1]
```diff
@@ -189,3 +189,14 @@ def create_systers_user(sender, instance, created, **kwargs):
...
+
+class UserSetting(models.Model):
+    user = models.OneToOneField(SystersUser, on_delete=models.CASCADE)
+    weekly_digest = models.BooleanField(default=False)
+    location_change = models.BooleanField(default=False)
+    time_change = models.BooleanField(default=False)
+    reminder = models.BooleanField(default=False)
+
+    def __str__(self):
+        return "Settings for {0}".format(self.user)
```
Comment (SanketDG) [inline systers_portal/users/models.py:211]: This seems incomplete to me. Do we not want something more verbose here?

File: systers_portal/users/apps.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from django.apps import AppConfig
+
+
+class UsersConfig(AppConfig):
+    name = 'users'
+
+    def ready(self):
+        import users.signals  # noqa # pylint: disable=unused-variable
```
Comment (SanketDG) [inline systers_portal/users/apps.py:8]: I don't seem to understand why this file exists.

File: systers_portal/users/forms.py [hunk 3]
```diff
@@ -51,3 +54,23 @@ def clean_password(self):
...
+        fields = ('weekly_digest', 'location_change', 'time_change', 'reminder')
+        helper_class = SubmitCancelFormHelper
+        helper_cancel_href = "{% url 'index' %}"
+
+    def __init__(self, *args, **kwargs):
+        self.user = kwargs.pop('user')
+        super(EditUserSettings, self).__init__(*args, **kwargs)
+
+    def save(self, commit=True):
+        """Override save to add created_by and meetup_location to the instance"""
```
Comment (SanketDG) [inline systers_portal/users/forms.py]: This docstring does not seem right.

File: systers_portal/users/views.py [hunk 4]
```diff
@@ -86,3 +88,23 @@ def check_permissions(self, request):
...
+
+class EditSettings(UpdateView, LoginRequiredMixin):
+    model = UserSetting
+    template_name = "users/settings.html"
+    form_class = EditUserSettings
+    raise_exception = True
+    form_valid_message = (u"Settings added Successfully")
+
+    def get_success_url(self):
+        return reverse('index')
```
Comment (SanketDG) [inline systers_portal/users/views.py]: Can we just reverse to the settings page?

File: systers_portal/users/views.py [hunk 5]
```diff
@@ -86,3 +88,23 @@ def check_permissions(self, request):
...
         permissions.extend([request.user == self.user])
         return any(permissions)
+
+
+class EditSettings(UpdateView, LoginRequiredMixin):
+    model = UserSetting
+    template_name = "users/settings.html"
+    form_class = EditUserSettings
+    raise_exception = True
+    form_valid_message = (u"Settings added Successfully")
```
Comment (SanketDG) [inline systers_portal/users/views.py]: no need of brackets

Also, this doesn't seem to appear when I update the settings.

**Review by SanketDG — CHANGES_REQUESTED** [review]

File: systers_portal/users/migrations/0002_usersetting.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    dependencies = [
+        ('users', '0001_squashed_0003_auto_20160207_1550'),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='UserSetting',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('weekly_digest', models.BooleanField(default=False)),
```
Comment (sakshi1499) [inline systers_portal/users/migrations/0002_usersetting.py]: I think Weekly digest should be true by default.

File: systers_portal/users/models.py (hunk 1 above)
Comment (sakshi1499) [reply systers_portal/users/models.py:211]: I also feel that as a user just having location change as an option wouldn't be very intuitive of what location change is referring.

File: systers_portal/users/views.py (hunk 4 above)
Comment (sakshi1499) [reply systers_portal/users/views.py]: +1

Comment (sakshi1499) [reply systers_portal/users/views.py]: But then the alert message should definitely work.

**Review by sakshi1499 — CHANGES_REQUESTED** [review]

File: systers_portal/users/models.py (hunk 1 above)
Comment (ritwickraj78) [reply systers_portal/users/models.py:211]: Sure I'll do that

File: systers_portal/users/apps.py (hunk 2 above)
Comment (ritwickraj78) [reply systers_portal/users/apps.py:8]: To trigger the signals when the app runs

File: systers_portal/users/forms.py (hunk 3 above)
Comment (ritwickraj78) [reply systers_portal/users/forms.py]: Yeah, I'll change it

File: systers_portal/users/views.py (hunk 4 above)
Comment (ritwickraj78) [reply systers_portal/users/views.py]: I'm not aware of why it doesnt work. I did not make any changes to it. Issue can be opened for this imo

Comment (ritwickraj78) [reply systers_portal/users/views.py]: @SanketDG Redirecting to the Profile Page would be a better option ig

File: systers_portal/users/migrations/0002_usersetting.py (hunk 6 above)
Comment (ritwickraj78) [reply systers_portal/users/migrations/0002_usersetting.py]: Sure, I'll do that

Comment (ritwickraj78) [reply systers_portal/users/migrations/0002_usersetting.py]: Done

File: systers_portal/users/views.py (hunk 5 above)
Comment (ritwickraj78) [reply systers_portal/users/views.py]: Brackets Removed

File: systers_portal/users/views.py (hunk 4 above)
Comment (ritwickraj78) [reply systers_portal/users/views.py]: Done

File: systers_portal/users/forms.py (hunk 3 above)
Comment (ritwickraj78) [reply systers_portal/users/forms.py]: Done

File: systers_portal/users/models.py (hunk 1 above)
Comment (ritwickraj78) [reply systers_portal/users/models.py:211]: Done

**Comment by ritwickraj78** [comment]
Screenshot from 2020-07-22 17-04-49

**Review by SanketDG — COMMENTED** [review]
Looks good to me!

I feel like you are lacking on tests though here. Will you do them later?

**Comment by ritwickraj78** [comment]
Created [thiis](https://github.com/anitab-org/portal/issues/609) issue and added GSoC Label. Will submit tests. Will that be good?

File: systers_portal/users/models.py [hunk 7]
```diff
@@ -189,3 +189,22 @@ def create_systers_user(sender, instance, created, **kwargs):
         if instance is not None:
             systers_user = SystersUser(user=instance)
             systers_user.save()
+
+
+class UserSetting(models.Model):
+    user = models.OneToOneField(SystersUser, on_delete=models.CASCADE)
+    weekly_digest = \
+        models.BooleanField(default=True,
```
Comment (SanketDG) [inline systers_portal/users/models.py]: This is weird formatting :smile:

File: systers_portal/users/models.py (hunk 1 above)
Comment (SanketDG) [reply systers_portal/users/models.py:211]: I don't see a change here

**Review by SanketDG — DISMISSED** [review]
This is almost done! :rocket:

File: systers_portal/users/models.py (hunk 7 above)
Comment (ritwickraj78) [reply systers_portal/users/models.py]: For the flake8 to pass. You have any suggestions on this?

Comment (SanketDG) [reply systers_portal/users/models.py]:
```suggestion
        models.BooleanField(
            default=True, verbose_name="Receive Weekly Digests from Communities")
```

Comment (ritwickraj78) [reply systers_portal/users/models.py]: @SanketDG Done

**Review by SanketDG — DISMISSED** [review]
This looks good to me!

**Review by sakshi1499 — DISMISSED** [review]
Awesome! See the failing checks though

**Review by sammy1997 — DISMISSED** [review]
Looks good. But can you change the commit messages so that the check passes? @ritwickraj78

**Review by SanketDG — APPROVED** [review]

**Pull request merged**